### ISSUE TITLE
Bug when installing (Energy) ADE (sequences)

### ISSUE THREAD

**Issue by gioagu** [3dcitydb/energy-ade-citydb#11]
Hi,

I've been able to reproduce a bug in the installation process of the Energy ADE (but I think it is in the ADE Manager plugin, actually). I'm using the 3DCityDB suite 2024.0.0 and the latest Energy ADE 2.10, and the 3dcitydb 4.4.1.
Besides the standard "citydb", I've created also another citydb schema, "citydb2". 
The search path variable is the standard one: search_path = citydb, citydb_pkg, public

Installation:

Case 1:
I want to install the Energy ADE on "citydb2" but not on "citydb". The installation will fail.

Case 2:
I first install the Energy ADE on "citydb", then I can also install it on "citydb2". here it will succeed, but there is a problem.

(I think) The problem is a bug in the way certain objects (sequences) are created and referenced in the SQL scripts. I think they are NOT schema qualified.
The installer looks for objects (e.g. sequences) in "citydb", and not in the actual installation schema ("citydb2"). So, in case 1, it will fail, but not in case 2. Sequences in "citydb" are used also for tables in "citydb2", which is wrong.

A negative side effect is that this also affects the uninstall process. If I try to first uninstall the ADE from "citydb", it will fail, as tables in "citydb2" make use of sequences in "citydb" instead of the ones in "citydb2".

Please get back to me if the description is not clear enough.
Feel free to move this issue to the 3dcitydb or somewhere else, if applicable.

Looking forward to the fixed version! ;-)

Best regards,

G

PS. Small screenshot as proof. Here "citydb2" is actually "alderaan".

![image](https://github.com/user-attachments/assets/5b955804-5c39-46dc-9a7b-947df69a3e6e)

After manually issuing

`ALTER TABLE alderaan.ng_dailyschedule ALTER COLUMN id SET DEFAULT nextval('alderaan.ng_dailyschedule_seq'::regclass);`

then the correct version is:

![image](https://github.com/user-attachments/assets/b5bd12de-8e5a-4d6c-9c09-df20eab4d661)

**Comment by clausnagel** [comment]
Thanks for reporting, @gioagu. Can you check this, @yaozhihang?

**Comment by yaozhihang** [comment]
thanks @gioagu, this is a bug in ADE manager plugin. I'll fix it this weekend.

**Comment by yaozhihang** [comment]
The issue could be resolved with the following changes in the importer/exporter code. It seems that we need to set the search path during the database connection to ensure that the DDL statements are executed against the correct target database schema.

``` java
public String getJDBCUrl(String server, int port, String database) {
     return "jdbc:postgresql://" + server + ":" + port + "/" + database + "?defaultRowFetchSize=10000
                 &reWriteBatchedInserts=true
                 &currentSchema=citydb2,citydb,public,citydb_pkg";
}
```

I also tested another method to set up the search path as shown below, but it didn't work.
``` java
statement.execute("SET search_path TO citydb2,citydb,public,citydb_pkg");
```

**Comment by clausnagel** [comment]
Hm, adding the following into the constructor of `PostgisADEDBSchemaManager` in the ADE Manager Plugin seemed to solve the issue for me:

```java
connection.prepareStatement("SET search_path TO " + schema + ", public").execute();
```

**Comment by gioagu** [comment]
Hi guys,
thanks for taking care of this bug! 👍

A small idea/suggestion: In the UNINSTALL DDL scripts of an ADE (the one stored in table ADE), would it be possible to add schema qualified names? This way, the SQL statements would work "out of the box", without need to (re)set the search_path every time.

Or, again, something at the beginning of the script (e.g. "SET search_path TO " + schema + ", public") that have the script work in the right schema, and maybe at the end to set the search_path back to the usual value?

Of course, just an idea, I didn't test it so far... ;-)

**Comment by clausnagel** [comment]
I guess, adding a `SET search_path TO my_schema, public` to the beginning of the script is possible, yes. But the uninstall scripts only delete the tables and sequences, not the database functions. So, would this really be helpful?

**Comment by gioagu** [comment]
Indeed, thanks for pointing it out. I agree with adding  `SET search_path TO my_schema, public` at the beginning of the script (But do no forget to set it back to `SET search_path TO citydb, citydb_pkg, public` at the end!

Besides, I'd have nothing against adding also the delete function statements, such as
`DELETE IF EXISTS my_schema.function(...) CASCADE;`
although, when everything works fine, you delete them programmatically during unregistration (if I have understood correctly). But if something goes wrong, you have a plan B. Anyway, just a wish... 🧑‍🎄

G

**Comment by yaozhihang** [comment]
I tested adding `SET search_path TO citydb2, public` at the beginning of the script. It doesn't work in my tests. So, the solution with `connection.prepareStatement("SET search_path TO " + schema + ", public").execute();` looks good to me.

Besides, adding schema-qualified names to the DDL scripts might reuqire some code changes. We could discuss it in a separate ticket.

**Comment by yaozhihang** [comment]
A fix [4fb7bae](https://github.com/3dcitydb/plugin-ade-manager/commit/4fb7baea0a609ad77f6d0d2c164d32ba23094bd7) is pushed to the release branch of the ADE manager

**Comment by clausnagel** [comment]
@gioagu, can you please confirm whether Zhihangs fix solves the problem for you?

**Comment by gioagu** [comment]
Perfect example of... lupus in fabula!!

I plan to test it in the days of this and next week! :-)

**Comment by RushiMUFC** [comment]
HI, @clausnagel @yaozhihang, I believe I will need a new jar file that reflects the above fixes, correct? Can you compile and release it as v2.1.1?

**Comment by clausnagel** [comment]
I'd prefer to have the test before we release. @RushiMUFC, you can easily clone the repository of the ADE Manager Plugin.

    > git clone https://github.com/3dcitydb/plugin-ade-manager.git

Afterwards, checkout the `release-2.3` branch which contains the fix.

    > git checkout release-2.3

Finally, you can create a build using Gradle (see also [here](https://github.com/3dcitydb/plugin-ade-manager/tree/release-2.3?tab=readme-ov-file#building)):

    > gradlew installDist

This build process will produce the plugin package under `build/install`. Simply copy the contents of this folder into the plugins folder of your Importer/Exporter installation to use the plugin.

**Comment by RushiMUFC** [comment]
Hi @clausnagel, thank you so much. I was able to generate a new build!

**Comment by clausnagel** [comment]
@RushiMUFC, nice. Can you confirm that the reported bug is fixed?

**Comment by gioagu** [comment]
Dear all,

I had time to test the latest changes. In short:
1) The issue with the sequences seems solved.  😊
If you have multiple citydb schemas, the tables in the second schema are correctly linked to the sequences of the second schema.

BUT... new bug 😒

2) Installation of an ADE on _another_ schema fails: ade delete functions are not installed (but the ade envelope functions are).

How to reproduce the bug:

- Install "citydb" as usual. Create additional "citydb2".
- Install Energy ADE into "citydb" AND into "citydb2"

Check the number of functions. In citydb2 all del_ng_...() functions are missing.
Do I have to open another ticket, or do I keep using this?

**Comment by yaozhihang** [comment]
Hm, thanks @gioagu for the feedback. I'll take a look the new bug.... You don't need to open another ticket.

**Comment by yaozhihang** [comment]
@gioagu could you test this branch?
https://github.com/3dcitydb/plugin-ade-manager/tree/hotfix-create-delete-scripts

**Comment by clausnagel** [comment]
Doesn't work for me, @yaozhihang. When I create a second schema "citydb2" and then register the ADE in "citydb2" and afterwards in "citydb", the number of database functions still differs also with your hotfix. I have 234 function in "citydb" and 254 in "citydb2". I am not sure whether this is systematic and deterministic, so that you will see the same difference on your machine.

**Comment by clausnagel** [comment]
Thanks, @yaozhihang. Your latest bugfix ([e1754d2d121c4db4aec070491a5ce05429fac8a0](https://github.com/3dcitydb/plugin-ade-manager/commit/e1754d2d121c4db4aec070491a5ce05429fac8a0)) solves all the issues raised in this thread in my tests.  

So, closing this thread now. @gioagu, @RushiMUFC, feel free to reopen if it doesn't work for you.
